src/app/admin/websiteDataTable/CustomEditComponent.tsx

```diff
@@ -0,0 +1,57 @@
+import { GridRenderEditCellParams, useGridApiContext } from "@mui/x-data-grid";
+import React, { useEffect, useLayoutEffect, useRef, useState } from "react";
+
+interface CustomCellProps extends GridRenderEditCellParams {
+    editMode: boolean;
+}
+
+const CustomComponent: React.FC<CustomCellProps> = (props: CustomCellProps) => {
+    const { id, field, hasFocus } = props;
+    const apiRef = useGridApiContext();
+    const ref = useRef<HTMLTextAreaElement>(null);
+    const initialValue = useRef(props.value);
+    const [value, setValue] = useState<string>(props.value);
+
+    useEffect(() => {
+        initialValue.current = props.value;
+        setValue(props.value);
+    }, [props.value]);
```

Comment:
so whenever the prop.value changes, this changes `initialValue` and `value`? Why can't the parent just manage the value and pass setValue down to this component?

Reply:
I thought about that, but I wasn't sure how we could keep track of all the states in the parent, given that we have an arbitrary number of rows in the grid, and we would need a state for each row. I'm sure that's possible but I don't know how to write that code

Reply:
Could the parent store some kind of array as a state? Just had a quick look around the code and the data grid docs. Could we do something like the following using generics so that the data grid is aware that `params.row` in each row is `WebsiteDataRow`?

```
// in WebsiteDataTable.tsx

export interface WebsiteDataRow {
    dbName: string;
    readableName: string;
    id: string;
    value: string;
}

...

const [rows, setRows] = useState<GridRowsProp<WebsiteDataRow>>([]);
...

const websiteDataColumns: GridColDef<WebsiteDataRow>[] = [
  ...,
  {
    renderCell: (params) => <EditableTextAreaForDataGrid {...params} editMode={false} value={params.row.value} />,
  },
  ...
]

...
```

Reply:
Ok. the types are working fine. What I've done is move the apiRef.current.setEditCellValue(...) call into the parent. I think this ref is holding all the values that we want, so I think this is the better way to do it.

No need for the value state in the child anymore, but we do still need the initialValue ref - this is purely to automatically set the cursor to the right position (the end) on clicking edit.